src/background/proxyHandler.js

```diff
@@ -0,0 +1,179 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+
+import { Component } from "./component.js";
+import { Logger } from "./logger.js";
+import { Utils } from "./utils.js";
+ 
+const log = Logger.logger("ProxyHandler");
+
+class SiteContext {
+  constructor(context) {
+    context = context || {};
+    this.origin = context.origin;
+    this.cityName = context.cityName;
+    this.countryCode = context.countryCode;
+    this.excluded = context.excluded || false;
+    this.proxyInfo = context.proxyInfo;
+  }
+}
+ 
+/**
+ * This class manages tasks related to creating and storing 
+ * proxy information (contexts) for specific origins. 
+ */
+export class ProxyHandler extends Component {
+  constructor(receiver) {
+    super(receiver); 
+  }
+
+  /**
+   * Creates a new SiteContext for an origin and adds 
+   * local socks proxy values for proxyInfo
+   * @param {string} origin - The origin to exclude.
+   */
+  async excludeOrigin(origin) {
+    const excluded = true;
+    // TODO: Replace with real local proxy info ✨
+    const proxyInfo = {
+      host: "us-dal-wg-socks5-504.relays.mullvad.net",
+      password: undefined,
+      port: "1080",
+      proxyDNS: true,
+      type: "socks",
+      username: undefined
+    };
+    const siteContext = new SiteContext({origin, excluded, proxyInfo});
+    return this.updateSiteContexts(siteContext);
+  }
+
+  /**
+   * Retrieves a Mullvad proxy url for a specified location.
+   * @param {string} countryCode - The two digit code for the country where the server is located.
+   * @param {string} cityName - The name of the city where the server is located.
+   */
+  async getMullvadProxy(countryCode, cityName) {
+    const proxyServer = this.pickServerBasedOnWeight(await this.getServersForLocation(countryCode, cityName));
+    return this.parseProxy(this.getMullvadProxyUrl(proxyServer.socksName));
+  }
+
+  /**
+   * Constructs a Mullvad proxy URL from a socks name.
+   * @param {string} socksName - The socks name.
+   */
+  getMullvadProxyUrl(socksName) {
+    return `socks://${socksName}.mullvad.net:1080`;
+  }
+
+  getRandomInteger(min, max) {
+    return Math.floor(Math.random() * (max - min + 1)) + min;
+  }
+
+  /**
+   * Retrieves an array of the Mullvad servers for a given city.
+   * @param {string} countryCode - The country code.
+   * @param {string} cityName - The city name.
+   */
+  async getServersForLocation(countryCode, cityName) {
+    const servers = await Utils.getServers();
```

Comment:
- Now this class has a direct dependency on web-ext storage. which makes it hard to unit-test.
- Utils now needs to consider how the vpn-controller stores things internally. 

This class has a dependency on a "thing" that provides it with a serverlist. 
*IMHO* We should express this in the constructor of the Component. i.e 

```js 
export class ProxyHandler extends Component {
  constructor(receiver, VPNController) {
  this.getServerList = ()=>{return VPNController.getState().serverlist}
 ```
and have main.js thread either pass VPNController or a func that will return the serverlist. 
This makes it easier to test.